### PR TITLE
Allow pollworkers to accumulate results from multiple print machines

### PR DESCRIPTION
**Pull request by carolinemodic**
Allows users to go 1-by-1 around print-enabled bmd machines (happy meal, or a print kiosk) and accumulate all of the results to one machine. 
https://zube.io/votingworks/vxsuite/c/3105 

Video

https://user-images.githubusercontent.com/14897017/112395399-84d11500-8cbb-11eb-997f-adf0468c826d.mov

Current Outputted report:
[output.pdf](https://github.com/votingworks/vxsuite/files/6200977/output.pdf)

Followups still todo after this PR
- Redesign the machine info in the PollsReport to not look ugly.
- Add undervotes and write in information to the PrecinctTallyReport (not strictly related to this feature but seems like something we should have)

### CODE REVIEW

File: apps/bmd/src/pages/PollWorkerScreen.tsx [hunk 1]
```diff
@@ -330,29 +466,70 @@ const PollWorkerScreen: React.FC<Props> = ({
             }
           />
         )}
+        {isSavingTally && (
+          <Modal content={<Loading>Saving to pollworker card…</Loading>} />
```
Comment (eventualbuddha) [inline apps/bmd/src/pages/PollWorkerScreen.tsx:470]: Either just "card" or "Poll Worker card" or "poll worker card" or something like that.
```suggestion
          <Modal content={<Loading>Saving to card…</Loading>} />
```

File: apps/bmd/src/components/PollsReport.tsx [hunk 2]
```diff
@@ -75,6 +80,7 @@ interface Props {
   isLiveMode: boolean
   isPollsOpen: boolean
   machineConfig: MachineConfig
+  machineMetadata?: CardTallyMetadataEntry[]
```
Comment (eventualbuddha) [inline apps/bmd/src/components/PollsReport.tsx:83]:
```suggestion
  machineMetadata?: readonly CardTallyMetadataEntry[]
```

File: apps/bmd/src/config/types.ts [hunk 3]
```diff
@@ -111,6 +111,17 @@ export type Tally = (
...
+export type CardTallyMetadataEntry = {
+  machineId: string
+  timeSaved: number
+}
+
+export type CardTally = {
+  tally: Tally
+  metadata: CardTallyMetadataEntry[]
+  totalBallotsPrinted: number
+}
```
Comment (eventualbuddha) [inline apps/bmd/src/config/types.ts:123]: Use `interface` and `readonly`?

File: apps/bmd/src/pages/PollWorkerScreen.test.tsx [hunk 4]
```diff
@@ -116,9 +130,265 @@ test('live mode on election day', async () => {
...
+      tally={getZeroTally(election)}
+      togglePollsOpen={jest.fn()}
+      saveTallyToCard={jest.fn()}
+      talliesOnCard={undefined}
+      clearTalliesOnCard={jest.fn()}
+    />
+  )
+
+  // eslint-disable-next-line no-restricted-syntax
+  expect(queryByText('Combine Results Reports')).toBe(null)
```
Comment (eventualbuddha) [inline apps/bmd/src/pages/PollWorkerScreen.test.tsx:168]: If you wanted to avoid the eslint disable directive you could do this:
```suggestion
  expect(queryByText('Combine Results Reports')).toBeNull()
```

File: apps/bmd/src/pages/PollWorkerScreen.tsx [hunk 5]
```diff
@@ -330,29 +466,70 @@ const PollWorkerScreen: React.FC<Props> = ({
...
+        )}
+        {isConfirmingCombinedPrint && (
+          <Modal
+            content={
+              <Prose>
+                <p>
+                  Do you want to print the combined results report from the{' '}
+                  {pluralize('machine', machineMetadata.length, true)} (
+                  {machineMetadata.map((m) => m.machineId).join(', ')}) and
+                  clear data from the card?
```
Comment (eventualbuddha) [inline apps/bmd/src/pages/PollWorkerScreen.tsx:483]: This reads slightly better to me:
```suggestion
                  erase the card?
```

File: apps/bmd/src/pages/PollWorkerScreen.tsx [hunk 6]
```diff
@@ -182,6 +219,81 @@ const PollWorkerScreen: React.FC<Props> = ({
...
+      timeSaved: Date.now(),
+    })
+    await saveTallyToCard({
+      tally: combinedTally,
+      metadata,
+      totalBallotsPrinted: combinedBallotsPrinted,
+    })
+    setIsSavingTally(false)
+  }
+  const metadataRows = []
```
Comment (eventualbuddha) [inline apps/bmd/src/pages/PollWorkerScreen.tsx:256]: Give this a type so it isn't inferred as `any[]`.
```suggestion
  const metadataRows: React.ReactChild[] = []
```

File: apps/bmd/src/pages/PollWorkerScreen.tsx [hunk 7]
```diff
@@ -330,29 +466,70 @@ const PollWorkerScreen: React.FC<Props> = ({
...
+              ballotsPrintedCount={
+                isPrintingCombinedReport
+                  ? combinedBallotsPrinted
+                  : ballotsPrintedCount
+              }
               currentDateTime={currentDateTime}
               election={election}
               isLiveMode={isLiveMode}
-              isPollsOpen={isPollsOpen}
+              isPollsOpen={isPrintingCombinedReport ? false : !isPollsOpen}
```
Comment (eventualbuddha) [inline apps/bmd/src/pages/PollWorkerScreen.tsx:514]: Why did this change to negate `isPollsOpen`? It reads like a mistake, but the logic is a little hard to follow so it's not clear whether it is or not.

File: apps/bmd/src/utils/Card.ts [hunk 8]
```diff
@@ -159,8 +159,7 @@ export class MemoryCard implements Card {
    */
   public async readLongObject<T>(): Promise<Optional<T>> {
     const { longValue } = this
-
-    if (!longValue) {
+    if (!longValue || longValue.length === 0) {
```
Comment (eventualbuddha) [inline apps/bmd/src/utils/Card.ts:162]: What triggered this change? Should we add a test for it?

File: apps/bmd/src/utils/tallies.test.ts [hunk 9]
```diff
@@ -192,3 +192,117 @@ test('tallies votes across many contests appropriately', () => {
...
+    candidates: [3, 0, 2, 1, 5, 0],
+    undervotes: 2,
+    writeIns: [
+      { name: 'eevee', tally: 2 },
+      { name: 'jolteon', tally: 3 },
+    ],
+  }
+
+  // Modify a yes no contest
+  tally1[12] = {
```
Comment (eventualbuddha) [inline apps/bmd/src/utils/tallies.test.ts:219]: These indexes feel fragile. Can we verify them or compute them?

File: apps/bmd/src/utils/tallies.ts [hunk 10]
```diff
@@ -6,10 +8,114 @@ import {
...
+  YesNoVoteTally,
+} from '../config/types'
 import { computeTallyForEitherNeitherContests } from './eitherNeither'
 import { getSingleYesNoVote } from './votes'
 
+const combineCandidateTallies = (
+  tally1: CandidateVoteTally,
+  tally2: CandidateVoteTally
+): CandidateVoteTally => {
+  const candidates = []
```
Comment (eventualbuddha) [inline apps/bmd/src/utils/tallies.ts:24]: Give it an annotation so it isn't inferred as `any[]`:
```suggestion
  const candidates: number[] = []
```

File: apps/bmd/src/utils/tallies.ts [hunk 11]
```diff
@@ -6,10 +8,114 @@ import {
...
+  assert(tally1.candidates.length === tally2.candidates.length)
```
Comment (eventualbuddha) [inline apps/bmd/src/utils/tallies.ts:25]: This would give you more information on a mismatch:
```suggestion
  assert.equal(tally1.candidates.length, tally2.candidates.length)
```

File: apps/bmd/src/utils/tallies.ts [hunk 12]
```diff
@@ -6,10 +8,114 @@ import {
...
+  for (let i = 0; i < tally1.candidates.length; i++) {
+    candidates.push(tally1.candidates[i] + tally2.candidates[i])
+  }
+  const writeInVotes = tally1.writeIns
+  for (const writeInVote of tally2.writeIns) {
+    const existingWriteInIdx = writeInVotes.findIndex(
+      (v) => v.name === writeInVote.name
+    )
+    if (existingWriteInIdx >= 0) {
+      writeInVotes[existingWriteInIdx].tally += writeInVote.tally
```
Comment (eventualbuddha) [inline apps/bmd/src/utils/tallies.ts:35]: This mutates `tally1`. I don't think it's clear from the function signature that it would do that. Can we make it not mutate its arguments?

File: apps/bmd/src/utils/tallies.ts [hunk 13]
```diff
@@ -6,10 +8,114 @@ import {
...
+}
+
+export const combineTallies = (
+  election: Election,
+  tally1: Tally,
+  tally2: Tally
+): Tally => {
+  assert(
+    election.contests.length === tally1.length &&
+      tally1.length === tally2.length
```
Comment (eventualbuddha) [inline apps/bmd/src/utils/tallies.ts:80]: I suggest using two assertions using `assert.equal` to get more information in the event of a crash.

File: apps/bmd/src/utils/tallies.ts [hunk 14]
```diff
@@ -6,10 +8,114 @@ import {
...
+  )
+  const combinedTally = [] as Tally
```
Comment (eventualbuddha) [inline apps/bmd/src/utils/tallies.ts:82]:
```suggestion
  const combinedTally: Tally = []
```

**Comment by carolinemodic** [comment]
Updated to address Brian's comments, and some small nits raised in show and tell 
- Removes double ellipsis
- Added ballot counts printed per machine to data model and table on the final report

File: apps/bmd/src/config/types.ts [hunk 15]
```diff
@@ -111,6 +111,18 @@ export type Tally = (
...
 
+export interface CardTallyMetadataEntry {
+  readonly machineId: string
+  readonly timeSaved: number
+  readonly ballotsPrinted: number
+}
+
+export interface CardTally {
+  readonly tally: Tally
+  readonly metadata: CardTallyMetadataEntry[]
```
Comment (eventualbuddha) [inline apps/bmd/src/config/types.ts:122]:
```suggestion
  readonly metadata: readonly CardTallyMetadataEntry[]
```

File: apps/bmd/src/utils/tallies.ts [hunk 16]
```diff
@@ -6,10 +8,112 @@ import {
...
+const combineCandidateTallies = (
+  tally1: CandidateVoteTally,
+  tally2: CandidateVoteTally
+): CandidateVoteTally => {
+  const candidates: number[] = []
+  assert.strictEqual(tally1.candidates.length, tally2.candidates.length)
+  for (let i = 0; i < tally1.candidates.length; i++) {
+    candidates.push(tally1.candidates[i] + tally2.candidates[i])
+  }
+  const writeInVotes = [...tally1.writeIns]
```
Comment (eventualbuddha) [inline apps/bmd/src/utils/tallies.ts:29]: This copies the array, but not its contents. So the mutation you have inside the loop is still mutating the original.

Comment (carolinemodic) [reply apps/bmd/src/utils/tallies.ts:29]: I'm actually refactoring how write ins work and remove this code in https://github.com/votingworks/vxsuite/pull/403

File: apps/bmd/src/pages/PollWorkerScreen.tsx (hunk 5 above)
Comment (benadida) [reply apps/bmd/src/pages/PollWorkerScreen.tsx:483]: "erase the card" might imply that it's no longer a poll worker card... could sound scarier. Maybe we should be even more specific, "clear tally data from the card"?

File: apps/bmd/src/pages/PollWorkerScreen.tsx (hunk 7 above)
Comment (benadida) [reply apps/bmd/src/pages/PollWorkerScreen.tsx:514]: agreed, I'm confused too.

File: apps/bmd/src/utils/Card.ts (hunk 8 above)
Comment (benadida) [reply apps/bmd/src/utils/Card.ts:162]: I suspect it's that we are clearing the card data on the poll worker card... and in the case of the voter card, where this also happens, the case has never come up because an expired voter card doesn't lead to reading the long value.

File: apps/bmd/src/utils/tallies.ts (hunk 16 above)
Comment (benadida) [reply apps/bmd/src/utils/tallies.ts:29]: just to be extra safe, I would suggest merging 403 into this first, so we can confirm this mutation no longer happens, before merging this to `main`. I'm being super careful here because we have been bitten before by mutating data structures, and so good to be extra extra careful.

**Review by benadida — CHANGES_REQUESTED** [review]
I agree with Brian's comments overall, and specifically on being careful with the mutating data structures, even if you're fixing them soon. Let's fix them first (maybe by merging the next PR into here), so `main` never has that behavior, just to be safe.

File: apps/bmd/src/pages/PollWorkerScreen.tsx (hunk 5 above)
Comment (carolinemodic) [reply apps/bmd/src/pages/PollWorkerScreen.tsx:483]: I like clear tally data.

File: apps/bmd/src/pages/PollWorkerScreen.tsx (hunk 7 above)
Comment (carolinemodic) [reply apps/bmd/src/pages/PollWorkerScreen.tsx:514]: So this report is printed right before the value of isPollsOpen is updated, currently it prints the correct report by having the code in PollsReport print a polls closed report when isPollOpen is true and a polls open report when isPollsOpen is false. I thought this was super confusing so I updated PollsReport to intepret isPollsOpen more "correctly" (true means print polls are open) and instead the confusion is moved into here. 

I agree its still confusing.... but I think this is a bit better then having the PollsReport component do something that seems backwords.

File: apps/bmd/src/utils/Card.ts (hunk 8 above)
Comment (carolinemodic) [reply apps/bmd/src/utils/Card.ts:162]: I was getting crazy explosions in integration tests before updating this (this is in MemoryCard, so this just impacts the test environment ) after clearing the long value, longValue here would be an empty UInt8Array and when you pass that to the TextDecoder below it exploded.

**Comment by benadida** [comment]
test response just to see how email responses work.

File: apps/bmd/src/AppRoot.tsx [hunk 17]
```diff
@@ -690,10 +724,16 @@ const AppRoot: React.FC<Props> = ({
           break
         }
         case 'pollworker': {
+          const isValid =
+            cardData.h === optionalElectionDefinition?.electionHash
+          const possibleCardTally: Optional<CardTally> = isValid
+            ? ((await card.readLongObject()) as Optional<CardTally>)
+            : undefined
```
Comment (benadida) [inline apps/bmd/src/AppRoot.tsx:731]: ah so this means that it will wait to declare a poll worker card is present until we read the long value. I can live with that for now, but it will mean that inserting the poll worker card will cause a longer lag before the screen reacts. Let's keep an eye on this when we test on live harware and see how this looks.

One tweak here: no reason to go read the long object if there is no long value stored on the card.

File: apps/bmd/src/pages/PollWorkerScreen.tsx [hunk 18]
```diff
@@ -182,6 +219,83 @@ const PollWorkerScreen: React.FC<Props> = ({
...
+      machineId: machineConfig.machineId,
+      timeSaved: Date.now(),
+      ballotsPrinted: ballotsPrintedCount,
+    })
+  }
+  if (talliesOnCard !== undefined) {
+    metadataRows.push(
+      ...talliesOnCard.metadata
+        .slice()
+        .sort((a, b) => b.timeSaved - a.timeSaved)
```
Comment (benadida) [inline apps/bmd/src/pages/PollWorkerScreen.tsx:283]: I believe this is a safe `sort` because of the `...` expansion, right? /cc @eventualbuddha

File: apps/bmd/src/pages/PollWorkerScreen.tsx (hunk 7 above)
Comment (benadida) [reply apps/bmd/src/pages/PollWorkerScreen.tsx:514]: I can live with this, but would recommend a comment explaining the meaning of `isPollsOpen` here.

**Review by benadida — APPROVED** [review]
Couple of small things to confirm / look at, otherwise good for me.

File: apps/bmd/src/AppRoot.tsx (hunk 17 above)
Comment (carolinemodic) [reply apps/bmd/src/AppRoot.tsx:731]: This seems to also be how we handle reading the long value on voter cards, in practice I haven't really noticed any major lag.

Comment (benadida) [reply apps/bmd/src/AppRoot.tsx:731]: good point. Maybe an optimization we can make later on both.

File: apps/bmd/src/AppRoot.tsx [hunk 19]
```diff
@@ -207,7 +210,11 @@ const initialAppState: Readonly<State> = {
 // Sets State. All side effects done outside: storage, fetching, etc
 type AppAction =
   | { type: 'processAdminCard'; electionHash: string }
-  | { type: 'processPollWorkerCard'; isPollWorkerCardValid: boolean }
+  | {
+      type: 'processPollWorkerCard'
+      isPollWorkerCardValid: boolean
+      talliesOnCard: Optional<CardTally>
```
Comment (eventualbuddha) [inline apps/bmd/src/AppRoot.tsx:216]: This would be more idiomatic:
```suggestion
      talliesOnCard?: CardTally
```

File: apps/bmd/src/AppRoot.tsx [hunk 20]
```diff
@@ -231,6 +238,10 @@ type AppAction =
   | { type: 'updateLastCardDataString'; currentCardDataString: string }
   | { type: 'activateCardlessBallot'; ballotStyleId: string }
   | { type: 'maintainCardlessBallot' }
+  | {
+      type: 'updatePollWorkerCardTally'
+      talliesOnCard: Optional<CardTally>
```
Comment (eventualbuddha) [inline apps/bmd/src/AppRoot.tsx:243]: Ditto:

```suggestion
      talliesOnCard?: CardTally
```

File: apps/bmd/src/AppRoot.tsx [hunk 21]
```diff
@@ -690,10 +724,20 @@ const AppRoot: React.FC<Props> = ({
           break
         }
         case 'pollworker': {
+          const isValid =
+            cardData.h === optionalElectionDefinition?.electionHash
+          /* istanbul ignore next */
+          // const shouldFetchLong = !!(isValid && longValueExists)
```
Comment (eventualbuddha) [inline apps/bmd/src/AppRoot.tsx:730]: Delete?

File: apps/bmd/src/pages/PollWorkerScreen.test.tsx [hunk 22]
```diff
@@ -116,9 +130,289 @@ test('live mode on election day', async () => {
...
+      printer={fakePrinter()}
+      tally={getZeroTally(election)}
+      togglePollsOpen={jest.fn()}
+      saveTallyToCard={jest.fn()}
+      talliesOnCard={undefined}
+      clearTalliesOnCard={jest.fn()}
+    />
+  )
+
+  // eslint-disable-next-line no-restricted-syntax
```
Comment (eventualbuddha) [inline apps/bmd/src/pages/PollWorkerScreen.test.tsx:167]: Comment outdated.
```suggestion
```

File: apps/bmd/src/utils/tallies.ts [hunk 23]
```diff
@@ -6,10 +8,104 @@ import {
...
+  tally1: CandidateVoteTally,
+  tally2: CandidateVoteTally
+): CandidateVoteTally => {
+  const candidates: number[] = []
+  assert.strictEqual(tally1.candidates.length, tally2.candidates.length)
+  for (let i = 0; i < tally1.candidates.length; i++) {
+    candidates.push(tally1.candidates[i] + tally2.candidates[i])
+  }
+  return {
+    candidates,
```
Comment (eventualbuddha) [inline apps/bmd/src/utils/tallies.ts:30]: Someday:
```suggestion
    candidates: zip(tally1.candidates, tally2.candidates, (c1, c2) => c1 + c2),
```

File: apps/bmd/src/pages/PollWorkerScreen.tsx (hunk 18 above)
Comment (eventualbuddha) [reply apps/bmd/src/pages/PollWorkerScreen.tsx:283]: It's safe because of the `.slice()`, which makes a copy. Another way would have been `[...talliesOnCard.metadata].sort()`, which is probably what you're thinking of @benadida.

**Review by eventualbuddha — APPROVED** [review]

File: apps/bmd/src/AppRoot.tsx (hunk 21 above)
Comment (carolinemodic) [reply apps/bmd/src/AppRoot.tsx:730]: lol oops
